### PR TITLE
geom/luks: LUKS v1 read-write support

### PR DESCRIPTION
**Pull request by n-p-soft**
This adds a new GEOM for LUKS (Linux Unified Key Support) disk encryption version 1.

### CODE REVIEW

**Comment by emaste** [comment]
We'd want to squash the followup commit (changing the debug level default) into the first commit. You can do that locally and push an update to this pull request, or it can be done by whoever merges this into the tree.

I apologize in advance that this will likely take some time to review (as would anything involving cryptography).

I would be interested in hearing a bit more about the background/context for this change, out of curiosity.

File: lib/geom/luks/Makefile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# $FreeBSD$
```
Comment (grahamperrin) [inline lib/geom/luks/Makefile:1]: [HEADS UP: $FreeBSD$ Removed from main](https://lists.freebsd.org/archives/freebsd-current/2023-August/004314.html)

**Comment by n-p-soft** [comment]
Roughly speaking, GELI = Linux dm-crypt (disk encryption, LUKS is the keystore) + Linux dm-verity (data integrity).When I planned to use FreeBSD as my main OS, I had two questions:1) is there a mean to access my Luks-encrypted data without doing a full decryption of the disks, and,2) will I need to go back to Linux in case some software is not running under FreeBSD ? As an example, I do not know yet if the AMD-Xilinx FPGA tools are natively supported using the Linux compatibility layer.In the first case, write support is not required. In the second, as well in the virtualization use case, write support is a nice mean to exchange encrypted data between Linux and FreeBSD.I will do some code changes according to all remarks, but I decided for now to make a try about Luks v2 support, using the v1 code as base, as only the Luks header is different in v2, but more annoying to parse. Tests are also needed for the write part.I sent the version 1 to get feedback and because the data I'm using daily is encrypted using Luks (currently, I'm running both Linux and FreeBSD).Le 3 oct. 2023 20:06, Mark Johnston ***@***.***> a écrit :
@markjdb commented on this pull request.

I left a few comments for cosmetic issues. A few larger comments:
How does this relate to GELI? Is your stance that this GEOM is solely to provide compatibility with existing LUKS volumes and that GELI should be used instead when possible? If so, the documentation should clearly state this. And, that being the case, do we really need write support?There are no regression tests. See tests/sys/geom/class for examples. I don't think we should add a new GEOM class without them.There are some style issues in the code. Please try running tools/build/checkstyle9.pl on the new files and fix warnings.

In lib/geom/luks/geom_luks.c:
> @@ -0,0 +1,72 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
+ *
+ * Copyright (c) 2023 Nicolas Provost ***@***.***>
+ * All rights reserved.

We generally avoid adding "All rights reserved" to new files: https://docs.freebsd.org/en/articles/committers-guide/#pref-license



In lib/geom/luks/geom_luks.c:
> + *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>

No need for cdefs.h here I think?



In lib/geom/luks/geom_luks.c:
> + * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <libgeom.h>
+
+#include "core/geom.h"
+
+

Extra newline.



In lib/geom/luks/gluks.8:
> +.\"    notice, this list of conditions and the following disclaimer in the
+.\"    documentation and/or other materials provided with the distribution.
+.\"
+.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
+.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
+.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+.\" SUCH DAMAGE.
+.\"
+.\" $FreeBSD$

No "$FreeBSD$" in new files, please.



In lib/geom/luks/gluks.8:
> +.Ar scheme
+.Ar pass
+.Nm
+.Cm info
+.Op Fl q
+.Ar device
+.Nm
+.Cm close
+.Ar provider
+.Sh DESCRIPTION
+The
+.Nm
+utility (e.g.
+.Xr geom 8 luks)
+is used for opening and creating LUKS (Linux Unified Key Setup)
+devices. For now, only the version 1 of LUKS is supported together with the

New sentences should begin on new lines.



In sys/geom/luks/g_luks.c:
> +
+	return (kproc_create(g_luks_worker, sc, &sc->worker, 0, 0,
+				"%s", sc->geom->name));
+}
+
+/************
+ * OP FUNCS *
+ ************/
+static inline size_t g_luks_roundss(size_t len, size_t sectorsize)
+{
+	size_t sslen = len / sectorsize;
+
+	if (len % sectorsize)
+		sslen++;
+
+	return (sslen * sectorsize);

This function can be written return (howmany(len, sectorsize));.

**Comment by bsdimp** [comment]
Looks like the review comments haven't been addressed, unless I'm missing something. Tagged this with 'changes required' until that happens. Also, using this a 'ping' to see if it's still relevant, or if it's gone in as a port, etc.

**Comment by bsdimp** [comment]
@n-p-soft This looks interesting... Do you plan on abandoning this for v2 support? Or is it still relevant with your plans? If it's going to be a while, I'd like to close this out, but invite you to reopen / create a new pull request when the time comes. What's up?

**Comment by bsdimp** [comment]
I'm going to close this as inactive, with an unclear path forward. If you implement v2 or feel that v1 has improved enough to include in FreeBSD, please open a new pull request. Thanks!

**Pull request closed without merging**